File: NSFEM.py
# translated from BASIC code in Ohnishi, et al.,
# "Flow Analysis by a personal computer" chapter 5 (in Japanese)

import numpy as np
from scipy.linalg import solveh_banded,norm
import matplotlib.pyplot as plt

def fix_bc(A,B,node,val):
    """ take care of dirichlet boundary condition """
    for n,v in zip(node,val):
        B[:] -= A[:,n]*v; B[n] = v
        A[:,n] = 0; A[n,:] = 0; A[n,n] = 1

def solveh_banded_(A,B,w):
    """ helper function for solving banded matrix eq """
    n = len(B)
    AB = np.empty((w+1,n))
    for j in range(w+1):# construct banded matrix
        AB[w-j,j:] = A[np.r_[j:n], np.r_[:n-j]]
    return solveh_banded(AB,B)

class NSFEM:
    """ Finite Element Method for solving
    Navier-Stokes equation in two dimensions.
    """
    def __init__(self, node, element,
                 stream_fix = None,
                 vortex_fix = None,
                 vortex_wall = None,
                 stream_free = None,
                 vortex_free = None,
                 visc=0, scale=1, offset=0):
        """
        node : string, file name of node data
          column 1 = x-coordinate of nodes
          column 2 = y-corrdinate of nodes
          column 3 = initial values of vortex
          if column3 is absent, initial values are all set to 0
        element : string, file name of element data
          three columns are node indices of three nodes in elements
          node indices begin from 0
        stream_fix : string
          file name of dirichlet boundary condition for stream function
          column 1 = indices of nodes on boundary
          column 2 = value of stream function on boundary
        vortex_fix : string
          file name of dirichlet boundary condition for vortex
          column 1 = indices of nodes on boundary
          column 2 = value of vortex on boundary
        vortex_wall : string
          file name of wall boundary condition for vortex
          column 1 = indices of nodes on boundary
          column 2,3 = indices of adjacent nodes
          if column 3 < 0, only column 2 is used
          else, mid point of 2 and 3 are used as adjacent node
          column 4 = wall slipping velocity (rightward from inside)
        stream_free : string
          file name of neumann boundary condition for stream function
          column 1 = indices of nodes on boundary
          column 2 = derivative of stream function on boundary
        vortex_free : string
          file name of neumann boundary condition for vortex
          column 1 = indices of nodes on boundary
          column 2 = derivative of vortex on boundary
        visc : viscousity coefficient
          used only if vortex_free is not None
        scale : scale factor to be muliplied to xy-coordinates
        offset : integer to be subtracted from node indices
        """
        # read data file
        node = np.loadtxt(node)
        nn = len(node) # number of nodes
        if node.shape[1] > 2:
            self.vortex = node[:,2] # initial values of vortex
            node = node[:,:2]
        else: self.vortex = np.zeros(nn)
        node *= scale

        elem = np.loadtxt(element).astype(np.int) - offset

        if stream_fix is None:
            self.ps_bc1_node, self.ps_bc1_val = [],[]
        else:
            a = np.loadtxt(stream_fix)
            self.ps_bc1_node = a[:,0].astype(np.int) - offset
            self.ps_bc1_val = a[:,1]

        if vortex_fix is None:
            self.om_bc1_node, self.om_bc1_val = [],[]
        else:
            a = np.loadtxt(vortex_fix)
            self.om_bc1_node = a[:,0].astype(np.int) - offset
            self.om_bc1_val = a[:,1]

        if vortex_wall is None:
            self.om_bc3_node, self.om_bc3_vs = [],[]
        else:
            a = np.loadtxt(vortex_wall)
            self.om_bc3_node = a[:,:3].astype(np.int) - offset
            self.om_bc3_vs = a[:,3] # slipping velocity

        if stream_free is None:
            self.ps_bc2_node, self.ps_bc2_val = [],[]
        else:
            a = np.loadtxt(stream_free)
            self.ps_bc2_node = a[:,:2].astype(np.int) - offset
            self.ps_bc3_val = a[:,2]

        if vortex_free is None:
            self.om_bc2_node, self.om_bc2_val = [],[]
        else:
            a = np.loadtxt(vortex_free)
            self.om_bc2_node = a[:,:2].astype(np.int) - offset
            self.om_bc2_val = a[:,2]

        # width of band matrix
        self.w = np.max(np.abs(np.diff(np.c_[elem, elem[:,0]])))

        # det = 2*(area of element) 
        xy = node[elem]
        dxy = np.roll(xy, -1, axis=1) - np.roll(xy, 1, axis=1)
        det = np.einsum('ij,ij->i', xy[:,:,0], dxy[:,:,1])

        # D = laplacian matrix, M = mass matrix
        cb = np.einsum('kij,k->kij', -dxy, 1/det) # c and -b
        bc = np.einsum('kil,kjl,k->kijl', cb, cb, det)/2
        D = np.sum(bc, axis=-1)
        M = (np.ones((3,3)) + np.eye(3))/24
        self.D = np.zeros((nn,nn))
        self.M = np.zeros((nn,nn))
        for k,e in enumerate(elem):
            i,j = np.meshgrid(e,e,indexing='ij')
            self.D[i,j] += D[k]
            self.M[i,j] += det[k]*M

        # distance between nodes for wall boundary condition
        self.om_bc3_dist = np.empty(len(self.om_bc3_node))
        for l,(i,j,k) in enumerate(self.om_bc3_node):
            if k<0: d = norm(node[j] - node[i])
            else: d = norm((node[j] + node[k])/2 - node[i])
            self.om_bc3_dist[l] = d

        # neumann boundary condition for stream function
        self.G2 = np.zeros(nn)
        for n,v in zip(self.ps_bc2_node, self.ps_bc2_val):
            self.G2[n] = v*norm(np.diff(node[n]))/2

        # neumann boundary condition for vortex
        self.G4 = np.zeros(nn)
        for n,v in zip(self.om_bc2_node, self.om_bc2_val):
            self.G4[n] = visc*v*norm(np.diff(node[n]))/2

        self.node = node
        self.element = elem
        self.cb = cb
        self.bc = np.flip(bc, axis=-1) # bb and cc
        self.visc = visc
        
    def run(self, dt, visc=None, eps=0, maxit=100, om=None):
        """
        dt : time step
        visc : viscousity coefficient (= 1/Re)
          if None, visc in __init__() is used
        eps : convergence criterion such that
          if relative change in max(abs(vortex)) < eps then exit
        maxit : maximum number of iterations
        om : initial values of vortex on nodes
          if None, initial values are all set to 0
          if scalar, initial values are all set to om
        return number of iterations
        if "ValueError: array must not contain infs or NaNs"
          occurs, try again
        """
        if visc is None: visc = self.visc
        nn = len(self.node) # number of nodes
        nw = len(self.om_bc3_node) # number of wall boundary conditions
        om_b = np.empty(nw) # vortex at wall boundary
        if om is None: om = self.vortex # vortex
        elif np.isscalar(om): om = np.full(nn, om)
        Mdt = self.M/dt

        for n in range(maxit):
            # poisson equation for stream function
            A = self.D.copy() # LHS
            B = np.dot(self.M, om) + self.G2 # RHS

            # dirichlet boundary condition
            fix_bc(A, B, self.ps_bc1_node, self.ps_bc1_val)

            # stream function
            ps = solveh_banded_(A, B, self.w)

            # compute velocity field (u,v)
            pe = ps[self.element]
            uv = np.einsum('ijk,ij->ik', self.cb, pe)

            # diffusion and convection terms
            A_ = visc + uv**2*dt/2 # numerical viscousity
            A_ = np.einsum('kl,kijl->kij', A_, self.bc)
            B_ = (np.roll(pe, -1, axis=-1) - np.roll(pe, 1, axis=-1))/6

            # vortex transport equation
            A = Mdt.copy()
            B = Mdt.copy()
            for k,e in enumerate(self.element):
                i,j = np.meshgrid(e,e,indexing='ij')
                A[i,j] += A_[k] # LHS
                B[i,j] -= B_[k]
    
            B = np.dot(B, om) + self.G4 # RHS

            # dirichlet boundary condition
            fix_bc(A, B, self.om_bc1_node, self.om_bc1_val)

            # wall boundary condition
            for l,(i,j,k) in enumerate(self.om_bc3_node):
                d = self.om_bc3_dist[l]
                vs = self.om_bc3_vs[l] # slipping velocity
                if k<0: ps_n,om_n = ps[j], om[j]
                else:
                    ps_n = (ps[j] + ps[k])/2
                    om_n = (om[j] + om[k])/2

                om_b[l] = 3*((ps[i] - ps_n)/d - vs)/d - om_n/2

            if nw: fix_bc(A, B, self.om_bc3_node[:,0], om_b)

            # votex
            om_ = om
            om = solveh_banded_(A, B, self.w)
            if np.max(np.abs(om-om_)) < eps*np.max(np.abs(om_)):
                break # converged

        self.stream = ps
        self.vortex = om
        self.velocity = uv
        return n+1

    def plot_element(self, *a, **k):
        x,y = self.node.T
        for e in self.element:
            e = np.r_[e,e[0]]
            plt.plot(x[e], y[e], *a, **k)

    def plot_velocity(self, *a, **k):
        x,y = np.sum(self.node[self.element], axis=1).T/3
        u,v = self.velocity.T
        plt.quiver(x,y,u,v, *a, **k)

    def save(self, node, element,
             stream_fix = None,
             vortex_fix = None,
             vortex_wall = None,
             stream_free = None,
             vortex_free = None,
             save_vortex = True):
        """ if save_vortex is True,
        vortex is saved in column 3 of node file """
        if save_vortex:
            np.savetxt(node, np.c_[self.node,
                                   self.vortex], '%g %g %g')
        else: np.savetxt(node, self.node, '%g %g')

        np.savetxt(element, self.element, '%d %d %d')

        if stream_fix is not None:
            np.savetxt(stream_fix,
                       np.c_[self.ps_bc1_node,
                             self.ps_bc1_val], '%d %g')
        if vortex_fix is not None:
            np.savetxt(vortex_fix,
                       np.c_[self.om_bc1_node,
                             self.om_bc1_val], '%d %g')
        if vortex_wall is not None:
            np.savetxt(vortex_wall,
                       np.c_[self.om_bc3_node,
                             self.om_bc3_vs], '%d %d %d %g')
        if stream_free is not None:
            np.savetxt(stream_free,
                       np.c_[self.ps_bc2_node,
                             self.ps_bc2_val], '%d %g')
        if vortex_free is not None:
            np.savetxt(vortex_free,
                       np.c_[self.om_bc2_node,
                             self.om_bc2_val], '%d %g')

File: C++/NSFEM.cpp
// Finite Element Method for solving
//   Navier-Stokes equation in two dimensions.
// translated from BASIC code in Ohnishi, et al.,
//  "Flow Analysis by a personal computer" chapter 5 (in Japanese)

#include<fstream>
#include<cstdlib>
#include<cmath>
#include "NSFEM.h"
#include "MatBandSym.h"
#include "util.h"
using namespace NSFEM;

mat_double NSFEM::node; // xy-coordinates of nodes
mat_int NSFEM::element; // node indices of elements
vec_double NSFEM::stream; // stream function at nodes
vec_double NSFEM::vortex; // vortex at nodes
mat_double NSFEM::velocity; // velocity at elements

static vec_int ps_bc1_node;// dirichlet boundary nodes for stream function
static vec_int om_bc1_node;// dirichlet boundary nodes for vortex
static mat_int om_bc3_node;// wall boundary nodes for vortex
static vec_int om_bc3_node0;// wall boundary nodes for vortex
static vec_double ps_bc1_val;// dirichlet boundary values for stream function
static vec_double om_bc1_val;// dirichlet boundary values for vortex
static vec_double om_bc3_vs; // slipping velocity
static vec_double om_bc3_dist;// distance between nodes
static vec_double G2;// neumann boundary condition for stream function
static vec_double G4;// neumann boundary condition for vortex
static vec_double det;// twice area of elements

static int nn; // number of nodes
static int ne; // number of elements;
static int nw; // number of wall boundaries
static int bw; // half width of banded matrix
static double visc; // viscosity
static MatBandSym D; // laplacian matrix
static MatBandSym M; // mass matrix

#define roll1(i) (i<2 ? i+1:0)
#define roll2(i) (i>0 ? i-1:2)

static void bce(double *b, double *c, const int *e)
// b = difference of y-coordinates of nodes in element e
// c = difference of x-coordinates of nodes in element e
{
    int i,j,k;
    for(i=0; i<3; i++) {
        j = e[roll1(i)];
        k = e[roll2(i)];
        b[i] = node[j][1] - node[k][1];
        c[i] = node[k][0] - node[j][0];
    }
}

static double dist(int i, int j)
// distance between nodes i and j
{
    return sqrt(pow(node[i][0] - node[j][0], 2)
              + pow(node[i][1] - node[j][1], 2));
}
    
static double dist(int i, int j, int k)
// distance between node i and midpoint of nodes j and k
{
    double d;
    d = pow(node[i][0] - (node[j][0] + node[k][0])/2, 2);
    d+= pow(node[i][1] - (node[j][1] + node[k][1])/2, 2);
    return sqrt(d);
}

static void fix_bc(MatBandSym& A, vec_double& B,
                   const vec_int& node, const vec_double& val)
// modify A and B to take care of
//   dirichlet boundary condition in (node, val)
{
    int i,j,k,l,n,m(B.size());
    for(n=0; n<node.size(); n++) {
        j = node[n];
        k = max(j-bw, 0);
        l = min(j+bw+1, m);
        for(i=k; i<l; i++) B[i] -= A(i,j)*val[n];
        for(i=k; i<l; i++) A(i,j) = 0;
        A(j,j) = 1;
        B[j] = val[n];
    }
}

static void fix_bc(vec_double& B,
                   const vec_int& node, const vec_double& val)
// modify B to take care of
//   dirichlet boundary condition in (node, val)
{
    for(int i=0; i<node.size(); i++)
        B[node[i]] = val[i];
}

void loadtxt(mat_double& A, std::ifstream& f);

void NSFEM::init(const char *node_file,
                 const char *element_file,
                 const char *stream_fix,
                 const char *vortex_fix,
                 const char *vortex_wall,
                 const char *stream_free,
                 const char *vortex_free,
                 double visc_, double scale, int offset)
// initializer to read data files
{
    int i,j,k,n[3],*e;
    double d;
    mat_double A;
    std::ifstream f;

    // read node data
    f.open(node_file);
    loadtxt(A,f);
    f.close();
    nn = A.nrows(); // number of nodes
    node.resize(nn, 2);
    for(i=0; i<nn; i++)
        for(j=0; j<2; j++)
            node[i][j] = A[i][j]*scale;

    vortex.assign(nn, 0.);
    if(A.ncols()>2)// initial values of vortex
        for(i=0; i<nn; i++) vortex[i] = A[i][2];

    // read element data
    f.open(element_file);
    loadtxt(A,f);
    f.close();
    ne = A.nrows(); // number of elements
    element.resize(ne, 3);
    for(i=0; i<ne; i++)
        for(j=0; j<3; j++)
            element[i][j] = int(A[i][j]) - offset;

    velocity.resize(ne, 2);

    // read dirichlet boundary data for stream function
    f.open(stream_fix);
    loadtxt(A,f);
    f.close();
    ps_bc1_node.resize(A.nrows());
    ps_bc1_val.resize(A.nrows());
    for(i=0; i<A.nrows(); i++)
        ps_bc1_node[i] = int(A[i][0]) - offset;
    for(i=0; i<A.nrows(); i++)
        ps_bc1_val[i] = A[i][1];

    // read dirichlet boundary data for vortex
    f.open(vortex_fix);
    loadtxt(A,f);
    f.close();
    om_bc1_node.resize(A.nrows());
    om_bc1_val.resize(A.nrows());
    for(i=0; i<A.nrows(); i++)
        om_bc1_node[i] = int(A[i][0]) - offset;
    for(i=0; i<A.nrows(); i++)
        om_bc1_val[i] = A[i][1];

    // read wall boundary data for vortex
    f.open(vortex_wall);
    loadtxt(A,f);
    f.close();
    nw = A.nrows(); // number of wall boundaries
    om_bc3_node.resize(nw, 3);
    om_bc3_node0.resize(nw);
    om_bc3_dist.resize(nw);
    om_bc3_vs.resize(nw);
    for(i=0; i<nw; i++) {
        for(j=0; j<3; j++)
            n[j] = om_bc3_node[i][j] = int(A[i][j]) - offset;
        om_bc3_node0[i] = n[0];
        om_bc3_vs[i] = A[i][3];
        if(n[2]<0) om_bc3_dist[i] = dist(n[0], n[1]);
        else om_bc3_dist[i] = dist(n[0], n[1], n[2]);
    }

    // width of banded matrix
    for(i=0; i<ne; i++) {
        e = element[i];
        bw = max(bw, abs(e[0]-e[1]));
        bw = max(bw, abs(e[1]-e[2]));
        bw = max(bw, abs(e[2]-e[0]));
    }

    // D = laplacian matrix, M = mass matrix
    double b[3], c[3];
    det.resize(ne);
    D.resize(nn, bw, 0.);
    M.resize(nn, bw, 0.);
    for(k=0; k<ne; k++) {
        e = element[k];
        bce(b,c,e);
        det[k] = b[0]*c[1] - b[1]*c[0];
        for(i=0; i<3; i++) {
            for(j=i; j<3; j++) {
                D(e[i],e[j]) += (b[i]*b[j] + c[i]*c[j])/det[k]/2;
                M(e[i],e[j]) += (i==j ? 2:1)/24.*det[k];
            }
        }
    }

    // read neumann boundary data for stream function
    f.open(stream_free);
    loadtxt(A,f);
    f.close();
    G2.assign(nn, 0.);
    for(i=0; i<A.nrows(); i++) {
        for(j=0; j<2; j++) n[j] = int(A[i][j]) - offset;
        d = dist(n[0],n[1]);
        G2[n[0]] = G2[n[1]] = A[i][2]*d/2;
    }
    
    // read neumann boundary data for vortex
    f.open(vortex_free);
    loadtxt(A,f);
    f.close();
    visc = visc_;
    G4.assign(nn, 0.);
    for(i=0; i<A.nrows(); i++) {
        for(j=0; j<2; j++) n[j] = int(A[i][j]) - offset;
        d = dist(n[0],n[1]);
        G4[n[0]] = G4[n[1]] = visc*A[i][2]*d/2;
    }

    // LDLT decomposition of the matrix D
    fix_bc(D, G2, ps_bc1_node, ps_bc1_val);
    LDLTdecomp(D,D);
}

int NSFEM::run(double dt, double visc_, double eps, int maxit)
// run simulation
{
    int i,j,k,l,m,n,*e;
    double b[3], c[3], u, v, d, ps_n, om_n;
    MatBandSym A, Mdt(M);
    MatBand C;
    vec_double B, om, om_b(nw);

    if(visc_ > 0) visc = visc_;
    for(i=0; i<M.nrows(); i++)
        for(j=0; j<M.ncols(); j++) Mdt[i][j] /= dt;
    for(m=0; m<maxit; m++) {// main loop
        // solve poisson equation for stream function
        B = G2;
        MulAddTo(B, M, vortex);
        fix_bc(B, ps_bc1_node, ps_bc1_val);
        solve(stream, D, B, false); // D is already decomposed
        // solve vortex transport equation
        A = Mdt;
        conv(C,A);
        for(n=0; n<ne; n++) {
            e = element[n];
            bce(b,c,e);
            // compute velocity field
            u = v = 0.;
            for(i=0; i<3; i++) u += c[i]*stream[e[i]];
            for(i=0; i<3; i++) v -= b[i]*stream[e[i]];
            velocity[n][0] = (u /= det[n]);
            velocity[n][1] = (v /= det[n]);
            // numerical viscousity
            u = (visc + u*u*dt/2)/det[n]/2;
            v = (visc + v*v*dt/2)/det[n]/2;
            // compute diffusion terms
            for(i=0; i<3; i++)
                for(j=i; j<3; j++)
                    A(e[i],e[j]) += u*b[i]*b[j] + v*c[i]*c[j];
            // compute convection terms
            for(j=0; j<3; j++) {
                k = e[roll1(j)];
                l = e[roll2(j)];
                d = (stream[k] - stream[l])/6;
                for(i=0; i<3; i++) C(e[i],e[j]) -= d;
            }
        }
        // wall boundary condition for vortex
        for(i=0; i<nw; i++) {
            e = om_bc3_node[i];
            d = om_bc3_dist[i];
            u = om_bc3_vs[i]; // slipping velocity
            if(e[2]<0) {
                ps_n = stream[e[1]];
                om_n = vortex[e[1]];
            }
            else {
                ps_n = (stream[e[1]] + stream[e[2]])/2;
                om_n = (vortex[e[1]] + vortex[e[2]])/2;
            }
            om_b[i] = 3*((stream[e[0]] - ps_n)/d - u)/d - om_n/2;
        }
        B = G4;
        MulAddTo(B, C, vortex);
        fix_bc(A, B, om_bc1_node, om_bc1_val);
        fix_bc(A, B, om_bc3_node0, om_b);
        om = vortex;
        solve(vortex, A, B);
        // check convergence
        u = v = 0.;
        for(i=0; i<nn; i++) {
            u = max(u, fabs(om[i] - vortex[i]));
            v = max(v, fabs(om[i]));
        }
        if(u < eps*v) return m+1;
    }
    return m;
}

void NSFEM::elem_center(double& x, double& y, int i)
// x,y = coordinates of mass center of element i
{
    const int *e(element[i]);
    x = (node[e[0]][0] + node[e[1]][0] + node[e[2]][0])/3;
    y = (node[e[0]][1] + node[e[1]][1] + node[e[2]][1])/3;
}